File: Python/Ex3.py
class HeapPriorityQueue:
    def __init__(self):
        self.heap = []

    def upheap(self, index):
        parent_index = (index - 1) // 2
        if index > 0 and self.heap[index] < self.heap[parent_index]:
            self.heap[index], self.heap[parent_index] = self.heap[parent_index], self.heap[index]
            self.upheap(parent_index)

    # Method to add an element, used for testing upheap
    def add(self, element):
        self.heap.append(element)
        self.upheap(len(self.heap) - 1)

# Testing the upheap method
heap = HeapPriorityQueue()
heap.add(10)
heap.add(4)
heap.add(15)
heap.add(3)  # Triggers upheap

print(heap.heap)

File: Java/Ex3.java
import java.util.ArrayList;

public class Ex3 {
    private ArrayList<Integer> heap;

    public Ex3() {
        heap = new ArrayList<>();
    }

    private void upheap(int index) {
        if (index == 0) {
            return;
        }

        int parentIndex = (index - 1) / 2;
        if (heap.get(index) < heap.get(parentIndex)) {
            int temp = heap.get(index);
            heap.set(index, heap.get(parentIndex));
            heap.set(parentIndex, temp);
            upheap(parentIndex);
        }
    }

    public void add(int element) {
        heap.add(element);
        upheap(heap.size() - 1);
    }

    public static void main(String[] args) {
        Ex3 pq = new Ex3();
        pq.add(10);
        pq.add(4);
        pq.add(15);
        pq.add(3); // Triggers upheap

        System.out.println(pq.heap);
    }
}
